### PR TITLE
[APM-788] [PRESUPP-555]   satellite joins for views  + terms filter for IN clauses

### PR DESCRIPTION
**Pull request by Dronplane**
### Scope & Purpose

Use satellite joins rule for views 
Optimize IN queries for SEARCH

Enterprise PR: https://github.com/arangodb/enterprise/pull/1252

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 1]
```diff
@@ -1549,7 +1693,11 @@ Result fromArrayComparison(irs::boolean_filter*& filter,
         // nothing to do more
         return {};
       }
-
+      TRI_ASSERT(!filter || filter->type() == irs::type<irs::And>::id() ||
+                 filter->type() == irs::type<irs::Or>::id());
+      byTermFilters.useByTerms =
+          byTermFilters.useByTerms && (matchCount == 1 || matchCount == n);
+      byTermFilters.allMatch = n == matchCount;
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1700]:
```suggestion
      byTermFilters.useByTerms &= (matchCount == n || matchCount == 1);
      byTermFilters.allMatch = matchCount == n;
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 2]
```diff
@@ -1462,6 +1595,15 @@ Result fromArrayComparison(irs::boolean_filter*& filter,
...
     toNormalize.reserve(2);
+    TRI_ASSERT(!filter || filter->type() == irs::type<irs::And>::id() ||
+               filter->type() == irs::type<irs::Or>::id());
+    // We can handle only 1 or all cases as due to mangling
+    // filters are distributed over several index fields. And
+    // for n-matches case we will need to support premutations, but in that case
+    // it is cheaper to run min-match disjunction
+    byTermFilters.useByTerms =
+        byTermFilters.useByTerms && (matchCount == 1 || matchCount == n);
+    byTermFilters.allMatch = n == matchCount;
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1606]:
```suggestion
    byTermFilters.useByTerms &= (matchCount == n || matchCount == 1);
    byTermFilters.allMatch = matchCount == n;
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 3]
```diff
@@ -1335,7 +1454,12 @@ class ByTermSubFilterFactory {
         termFilter = &appendNot<irs::by_term>(
             append<irs::And>(*filter, filterCtx), filterCtx);
       } else {
-        termFilter = &append<irs::by_term>(*filter, filterCtx);
+        if (termsOpt.useByTerms) {
+          return appendByTermsFilter(filter, value, fieldName, filterCtx,
+                                     termsOpt);
+        } else {
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1460]: else is unnecessary

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 4]
```diff
@@ -1316,7 +1416,25 @@ class ByTermSubFilterFactory {
...
+          }
+          ScopedAqlValue value(*node.value);
+          if (!value.isConstant()) {
+            if (!value.execute(filterCtx.query)) {
+              // failed to execute expression
+              return {TRI_ERROR_BAD_PARAMETER, "could not execute expression"};
+            }
+          }
+          return appendByTermsFilter(filter, value, name, filterCtx, termsOpt);
+        } else {
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1435]: else is unnecessary

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 5]
```diff
@@ -1437,7 +1569,8 @@ Result fromArrayComparison(irs::boolean_filter*& filter,
     }
     Result buildRes;
     aql::AstNodeType arrayExpansionNodeType;
-    std::tie(buildRes, arrayExpansionNodeType) =
+    size_t matchCount;
+    std::tie(buildRes, arrayExpansionNodeType, matchCount) =
```
Comment (jsteemann) [inline arangod/IResearch/IResearchFilterFactory.cpp:1573]: nit: could be made simpler by using `auto [buildRes, arrayExpansionNodeType, matchCount] = ...`
```suggestion
    std::tie(buildRes, arrayExpansionNodeType, matchCount) =
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 6]
```diff
@@ -1290,24 +1298,116 @@ std::pair<Result, aql::AstNodeType> buildBinaryArrayComparisonPreFilter(
...
+  containers::FlatHashMap<std::string, irs::by_terms*> filters;
+  bool useByTerms{false};
+  bool allMatch{false};
+};
+
+Result appendByTermsFilter(irs::boolean_filter* filter,
+                           ScopedAqlValue const& value, std::string& name,
+                           FilterContext const& filterCtx,
+                           ByTermsOptimizationContext& termsOpt) {
+  auto makeFilter = [&]() -> irs::by_terms* {
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1334]: Why don't return reference? In such case don't need to think about nullptr possibility on caller side

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 7]
```diff
@@ -1537,8 +1680,9 @@ Result fromArrayComparison(irs::boolean_filter*& filter,
     case SCOPED_VALUE_TYPE_ARRAY: {
       size_t const n = value.size();
       Result buildRes;
+      size_t matchCount;
       aql::AstNodeType arrayExpansionNodeType;
-      std::tie(buildRes, arrayExpansionNodeType) =
+      std::tie(buildRes, arrayExpansionNodeType, matchCount) =
```
Comment (jsteemann) [inline arangod/IResearch/IResearchFilterFactory.cpp:1685]: nit: could be simplified by using `auto [buildRes, arrayExpansionNodeType, matchCount] = ...`
```suggestion
      std::tie(buildRes, arrayExpansionNodeType, matchCount) =
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 8]
```diff
@@ -1290,24 +1298,116 @@ std::pair<Result, aql::AstNodeType> buildBinaryArrayComparisonPreFilter(
...
+    auto existing = termsOpt.filters.find(name);
+    if (existing == termsOpt.filters.end()) {
+      auto& terms_filter = filter->add<irs::by_terms>();
+      *terms_filter.mutable_field() = name;
+      terms_filter.boost(filterCtx.boost);
+      termsOpt.filters.emplace(name, &terms_filter);
+      return &terms_filter;
+    } else {
+      return existing->second;
+    }
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1344]:
```suggestion
    if (existing != termsOpt.filters.end()) {
      return existing->second;
    }
    auto& terms_filter = filter->add<irs::by_terms>();
    *terms_filter.mutable_field() = name;
    terms_filter.boost(filterCtx.boost);
    termsOpt.filters.emplace(name, &terms_filter);
    return &terms_filter;
```

File: tests/IResearch/IResearchFilterArrayInTest.cpp [hunk 9]
```diff
@@ -81,12 +81,169 @@
...
+using FilterIterator =
+    decltype(reinterpret_cast<irs::boolean_filter*>(0)->begin());
+
+void checkTermFilter(FilterIterator begin, bool){// 1st filter
+                                                 {irs::by_term expected;
+*expected.mutable_field() = mangleStringIdentity("a.b.c.e.f");
+expected.mutable_options()->term =
+    irs::ViewCast<irs::byte_type>(std::string_view("1"));
+EXPECT_EQ(expected, **begin);
+}  // namespace
```
Comment (jsteemann) [inline tests/IResearch/IResearchFilterArrayInTest.cpp:97]: nit: indentation here is somewhat broken.
I guess this is due to the comment at the beginning of the function body.

File: arangod/Aql/IResearchViewNode.cpp [hunk 10]
```diff
@@ -1459,6 +1459,22 @@ std::pair<bool, bool> IResearchViewNode::volatility(
...
 
+aql::Collection const* IResearchViewNode::collection() const {
+  auto c = collections();
+  if (c.empty()) {
+    return nullptr;
+  }
+  // this API should not be used for multicollection views
+  // doing so migth break satellite join assumptions
+  TRI_ASSERT(c.size() == 1);
+  return &c.front().first.get();
```
Comment (MBkkt) [inline arangod/Aql/IResearchViewNode.cpp:1470]:
```suggestion
  auto c = collections();
  if (c.size() != 1) {
    // this API should not be used for multicollection views
    // doing so migth break satellite join assumptions
    TRI_ASSERT(c.size() == 0);
    return nullptr;
  }
  return &c.front().first.get();
```

**Review by jsteemann — APPROVED** [review]
LGTM

File: tests/IResearch/IResearchFilterArrayInTest.cpp (hunk 9 above)
Comment (MBkkt) [reply tests/IResearch/IResearchFilterArrayInTest.cpp:97]: yes, `{ // comment ` formatted bad

File: arangod/IResearch/IResearchFilterFactory.cpp (hunk 4 above)
Comment (Dronplane) [reply arangod/IResearch/IResearchFilterFactory.cpp:1435]: Ah, I see, you mean because of return, not because of impossible to get here

**Review by MBkkt — APPROVED** [review]

**Review by MBkkt — APPROVED** [review]

**Pull request merged**